keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,951 @@
+  system.
+* It is especially difficult to know if use of a feature at some point makes
+  cluster upgrade unsafe now; for example, an alpha object being written while
+  the feature was briefly on which is incompatable with the default-on beta.
+* It is not easy to test the feature on -> use -> feature off lifecycle.
+
+### Goals
+
+Primary goals:
+* Minimize codebase churn, existing feature authors do little-to-no work
+* Expose the feature states in use in the cluster
+* Make it possible to get the cluster back into the "off" state
+* Expose the "off but dangling references / uses" state
+* Make it easy to add enablement and disablement code
+* Make it easy to add upgrade and downgrade code
+* Make it easy to tell if it is not safe to upgrade / downgrade due to feature
+  states
+* A test ensures that feature on -> use -> feature off works for all features
+  (i.e. it's easy for feature authors to be included in this test)
+
+Secondary goals:
+* Third parties should be able to use the API and libraries we build for their
+  own features.
+* Set feature states centrally (on apiserver command line) -- may not apply to
+  all features
+* Permit the platform admin to permit selected features to be dynamically
+  settable, by api clients
+
+Follow-up goals:
+* When a feature is about a new field, remove it from OpenAPI specs when the
+  feature is off.
+* For features that possibly could support it, add a canarying mechanism so that
+  the feature enablement doesn't break the whole cluster at once when there is a
+  problem.
+
+### Non-Goals
+
+* The feature system will not permit passing configuration, since that's already
+  done in a variety of ways today. Therefore only features that don't take
+  config can be dynamic. Maybe this can be addressed in a follow up.
+
+## Proposal
+
+### High-level design
+
+First we will discuss all the pieces with minimal detail. Probably this section
+will leave you with questions which hopefully are answered in the "detailed
+design" section.
+
+#### Client Flow
+
+Let's call the current use of feature gates the "client flow". Things that are
+already handled correctly by the existing client flow we will try not to change.
+
+The main thing we want to add to the client flow is reporting of the feature's
+uses: did the client check the feature state, and what value did it find? The
+necessity of this will be clear when we discuss the "server flow" later.
+
+To make our improvements to the client flow, we will add a new implementation of
+the existing k8s.io/component-base/featuregate package's MutableFeatureGate
+type. It will require post-construction initialization, so that it can write the
+uses to the api; to be a drop-in replacement, it can't require an api client
+when it is constructed, because that happens before the command line is parsed /
+config files are read.
+
+Our new client will use the existing MutableFeatureGate to read the command
```

Comment:
So the client is a binary, and using this to read its own command line for feature gate values? Or the client is the author of a feature that is included in the binary?

Reply:
Yeah, sorry, I wasn't clear either, for the latter I meant: "the code embedded in the binary, checking if the feature is on in order to do it's thing or not".

Reply:
Given that one binary supports many features, the parsing of the command line is code written only once by the binary author (or maybe automatic in a lib), whereas checking the feature gate is code every feature author will need to write. So, trying to understand which flow this is, and why we start with it. I would think we would start with how feature gates are consumed by feature authors in their feature code; that's the most common use.

Reply:
The way it works today is that there's a FeatureGate implementation already written for people, the binary declares it statically and installs its flags in the command line parser, and feature authors call Enabled(f) for each feature f. The only thing in this story that will change is where the FeatureGate implementation lives. This is all part of the client flow, it seems like you're asking which half of this I'm talking about, but I'm talking about all of it.

Reply:
Does the description of "clients" I added above fix this or is it still unclear?